cardano-crypto-class/src/Cardano/Crypto/Libsodium/Memory/Internal.hs

```diff
@@ -49,33 +125,115 @@ traceMLockedForeignPtr fptr = withMLockedForeignPtr fptr $ \ptr -> do
 
 {-# DEPRECATED traceMLockedForeignPtr "Don't leave traceMLockedForeignPtr in production" #-}
 
+makeMLockedPool :: forall n. KnownNat n => IO (Pool n)
+makeMLockedPool = do
+  initPool
+    (fromIntegral $ 4096 `div` (natVal (Proxy @n)) `div` 64)
+    (\size -> do
+      ptr <- sodiumMalloc (fromIntegral size)
+      newForeignPtr ptr (sodiumFree ptr)
+    )
+    (\ptr -> do
+      eraseMem (Proxy @n) ptr
+      pushAllocLogEvent $ FreeEv (ptrToWordPtr ptr)
+    )
+
+eraseMem :: forall n a. KnownNat n => Proxy n -> Ptr a -> IO ()
+eraseMem proxy ptr = fillBytes ptr 0xff (fromIntegral $ natVal proxy)
+
+mlpool32 :: Pool 32
+mlpool32 = unsafePerformIO makeMLockedPool
+{-#NOINLINE mlpool32 #-}
+
+mlpool64 :: Pool 64
+mlpool64 = unsafePerformIO makeMLockedPool
+{-#NOINLINE mlpool64 #-}
+
+
 -- | Allocate secure memory using 'c_sodium_malloc'.
 --
 -- <https://libsodium.gitbook.io/doc/memory_management>
 --
-allocMLockedForeignPtr :: Storable a => IO (MLockedForeignPtr a)
-allocMLockedForeignPtr = impl undefined where
-    impl :: forall b. Storable b => b -> IO (MLockedForeignPtr b)
-    impl b = do
-        ptr <- sodiumMalloc size
-        fmap SFP (newForeignPtr c_sodium_free_funptr ptr)
+-- allocMLockedForeignPtr :: Storable a => IO (MLockedForeignPtr a)
+-- allocMLockedForeignPtr = impl undefined where
+--     impl :: forall b. Storable b => b -> IO (MLockedForeignPtr b)
+--     impl b = do
+--         ptr <- sodiumMalloc size
+--         let finalizer = sodiumFree ptr
+--         fmap SFP (newForeignPtr ptr finalizer)
+--
+--       where
+--         size :: CSize
+--         size = fromIntegral size''
+--
+--         size' :: Int
+--         size' = sizeOf b
+--
+--         align :: Int
+--         align = alignment b
+--
+--         size'' :: Int
+--         size''
+--             | m == 0    = size'
+--             | otherwise = (q + 1) * align
+--           where
+--             (q,m) = size' `divMod` align
+
+allocMLockedForeignPtr :: forall a. Storable a => IO (MLockedForeignPtr a)
+allocMLockedForeignPtr = do
+  fptr <- mlockedMalloc size
+  withForeignPtr fptr $ \ptr -> pushAllocLogEvent $ AllocEv (ptrToWordPtr ptr)
+
+  return $ SFP . castForeignPtr $ fptr
```

Comment:
I can actually explain that.

It's a variation of the dot-dot-dollar pattern, where you apply a composed chain of functions to one final argument (`f . g . h . i . j $ x`); except that here, we fence off the `return` from the other chained functions, to create a bit of a grouping that highlights the parallels between a pure composition chain (`SFP . castForeignPtr $ fptr`) and the same chain wrapped in `return` to fit it into a monadic context (`return $ SFP . castForeignPtr $ fptr`). IMO this fits in nicely with the quasi-imperative feel of `do` notation.

I'm not super attached to this, if you prefer a different style, then I'll change it.

Reply:
I don't particularly care about this pattern: `f . g . h . i . j $ x`, if someone uses it then that's their prerogative. However, I don't get why would someone write: `i . j $ x` instead of `i $ j x` or `i (j x)`. Latter two require less typing and read easier. 

This is just mere discussion, so feel free to go about this either way you like it.

Reply:
As I said, not super opinionated, though I happen to like the dot-dot-dollar style.